bootstrap_conan_utils.py

```diff
@@ -0,0 +1,565 @@
+#!/usr/bin/env python
+
+from dataclasses import dataclass
+
+import json
+import os
+import platform
+import requests
+import sys
+import unittest
+import urllib.parse
+
+DEFAULT_CONAN_URL = "http://conan.jitx.com:8081/artifactory/api/conan/conan-local"
+
+def eprint(msg):
+    print(msg, file=sys.stderr)
+
+def debug(msg):
+    pass
+    eprint(msg)
+
+if sys.version_info[0] < 3:
+    eprint("This script requires python version 3 or greater")
+    exit(1)
+
+@dataclass
+class ConanVersion:
+    """
+    A structure for holding all of the components of a fully-qualified conan version
+
+    https://docs.conan.io/2/tutorial/versioning/revisions.html
+    """
+    name: str
+    version: str
+    recipe_revision: str = None
+    package_id: str = None
+    package_revision: str = None
+
+    @staticmethod
+    def from_string(s: str):
+        """
+        A constructor to parse a fully-qualified conan version string into a ConanVersion
+
+        Parameters:
+            s: str in the form "a/b#cccccccccccccccccccccccccccccccc:dddddddddddddddddddddddddddddddddddddddd#eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee"
+                   where only "a/b" is mandatory
+
+        Throws Exception If we fail to parse the version string
+        """
+        debug(f"ConanVersion: s==\"{s}\"")
+        # name before slash
+        pslash = s.partition('/')
+        name = pslash[0]
+        pcolon = pslash[2].partition(':')
+        pver = pcolon[0].partition('#')
+        ppkg = pcolon[2].partition('#')
+        version = pver[0]
+        recipe_rev = pver[2] if pver[2] else None
+        package_id = ppkg[0] if ppkg[0] else None
+        package_rev = ppkg[2] if ppkg[2] else None
+        debug(f"ConanVersion( \"{name}\", \"{version}\", \"{recipe_rev}\", \"{package_id}\", \"{package_rev}\")")
```

Comment:
Why not use `re`:

```python
regex = re.compile(r"^(?P<name>[^\/:#]+)\/(?P<version>[^\/:#]+)(#(?P<recipe_rev>[^\/:#]+)(?P<pkg_id>:([^\/:#]+)(#(?P<pkg_rev>.*))?)?)?$")
m = regex.match(s)
name = m.group("name")
version = m.group("version")
recipe_rev = m.group("recipe_rev")
package_id = m.group("pkg_id")
package_rev = m.group("pkg_rev")
...
```

Reply:
Maintainability?  That regex looks pretty hairy.

"Some people, when confronted with a problem, think "I know, I'll use regular expressions." Now they have two problems." - jwz, alt.religion.emacs